fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,336 @@
+#define FDBSERVER_TENANTENTRYCACHE_ACTOR_H
+
+#pragma once
+
+#include "fdbclient/DatabaseContext.h"
+#include "fdbclient/FDBOptions.g.h"
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/RunTransaction.actor.h"
+#include "fdbclient/Tenant.h"
+#include "fdbclient/TenantManagement.actor.h"
+#include "fdbserver/Knobs.h"
+#include "flow/IndexedSet.h"
+
+#include <functional>
+#include <unordered_map>
+
+#include "flow/actorcompiler.h" // has to be last include
+
+#define INITIAL_CACHE_GEN 1
+
+using TenantEntryCacheGen = uint64_t;
+using TenantNameEntryPair = std::pair<TenantName, TenantMapEntry>;
+using TenantNameEntryPairVec = std::vector<TenantNameEntryPair>;
+
+enum class TenantEntryCacheRefreshReason { INIT = 1, PERIODIC_TASK = 2, CACHE_MISS = 3 };
+
+template <class T>
+struct TenantEntryCachePayload {
+	TenantName name;
+	TenantMapEntry entry;
+	TenantEntryCacheGen gen;
+	// Custom client payload
+	T payload;
+};
+
+template <class T>
+using TenantEntryCachePayloadFunc =
+    std::function<TenantEntryCachePayload<T>(const TenantName&, const TenantMapEntry&, const TenantEntryCacheGen)>;
+
+// In-memory cache for TenantEntryMap objects. It supports three indices:
+// 1. Lookup by 'TenantId'
+// 2. Lookup by 'TenantPrefix'
+// 3. Lookup by 'TenantName'
+//
+// TODO:
+// ----
+// The cache allows user to construct the 'cached object' by supplying a callback to allow constructing the cached
+// object. The cache implements a periodic refresh mechanism, polling underlying database for updates (add/remove
+// tenants), in future we might want to implement database range-watch to monitor such updates
+
+template <class T>
+class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyable {
+private:
+	UID uid;
+	Database db;
+	TenantEntryCacheGen curGen;
+	TenantEntryCachePayloadFunc<T> createPayloadFunc;
+
+	AsyncTrigger sweepTrigger;
+	Future<Void> sweeper;
+	Future<Void> refresher;
+	Map<int64_t, TenantEntryCachePayload<T>> mapByTenantId;
+	Map<TenantName, TenantEntryCachePayload<T>> mapByTenantName;
+
+	// TODO: add hit/miss/refresh counters
+
+	ACTOR static Future<TenantNameEntryPairVec> getTenantList(Reference<ReadYourWritesTransaction> tr) {
+		tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+
+		KeyBackedRangeResult<TenantNameEntryPair> tenantList = wait(TenantMetadata::tenantMap.getRange(
+		    tr, Optional<TenantName>(), Optional<TenantName>(), CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER + 1));
+		ASSERT(tenantList.results.size() <= CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER && !tenantList.more);
+
+		TraceEvent(SevDebug, "TenantEntryCache_GetTenantList").detail("Count", tenantList.results.size());
+
+		return tenantList.results;
+	}
+
+	ACTOR static Future<Void> refreshImpl(TenantEntryCache<T>* cache, TenantEntryCacheRefreshReason reason) {
+		TraceEvent("TenantEntryCache_RefreshStart", cache->id()).detail("Reason", static_cast<int>(reason));
```

Comment:
I could see these being super verbose with many tenants, we might want to make these SevDebug and rely more on counters by reason to understand behavior

Reply:
Good point.I think it should be SevDebug with `suppression` param plugged in. Will do that. 
Thanks for pointing it out.